src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node head;
+    private Node tail;
+    private int size = 0;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(tail, value, null);
+        if (head == null) {
+            head = node;
+        }
+        if (tail != null) {
+            tail.next = node;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Can't add value to index "
+                    + index + ", list's last index is " + (size - 1));
+        }
+        Node<T> currentINode = getNode(index);
+        Node<T> newNode = new Node<>(currentINode.prev, value, currentINode);
+        if (currentINode.prev == null) {
+            head = newNode;
+        } else {
+            currentINode.prev.next = newNode;
+        }
+        currentINode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new RuntimeException("Can't add list, it is null");
+        }
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index "
+                    + index + ", last index = " + (size - 1));
+        }
+        Node<T> currentNode = getNode(index);
+        if (currentNode != null) {
+            return currentNode.value;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        Node<T> currentNode = getNode(index);
+        if (currentNode != null) {
+            T prevValue = currentNode.value;
+            currentNode.value = value;
+            return prevValue;
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = getNode(index);
+        return removeNode(node);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        while (node != null) {
+            if ((node.value == null && object == null) || node.value.equals(object)) {
```

Comment:
```suggestion
            if ((node.value == object) || node.value.equals(object)) {
```
другу частину після ||  можна ще покращити. Наразі там може бути NPE якщо node.value == null і object != null. Тоді перша частина дасть false а друга NPE бо на налі викличемо .equals()